[primTorch] Clean up `softmax` and `log_softmax` refs, add aliases, add `softmin` ref

nkaretnikov opened the pull request.

Commit:
Remove keyword-only argument from `softmax` ref

For compat with ATen and Python frontend signatures.  See the comment in
`native_functions.yaml`: "softmax allows positional dtype, unlike most
operators, because kwonly is BC-breaking when loading jit models."

Commit:
In `softmax` ref, use `dtype` in `get_computation_dtype`

Commit:
Make sure `softmax` decomposes into prim calls

```
import torch
from torch._prims.context import TorchRefsMode
from torch.fx.experimental.proxy_tensor import make_fx

a = torch.randn(3, 3, device="cuda", dtype=torch.float32)

with TorchRefsMode(strict=True):
    make_fx(lambda x: torch.softmax(x, 1))(a).graph.print_tabular()
```

Output:

```
opcode         name                    target                              args                                  kwargs
-------------  ----------------------  ----------------------------------  ------------------------------------  --------
placeholder    x_1                     x_1                                 ()                                    {}
call_function  convert_element_type    prims.convert_element_type.default  (x_1, torch.float32)                  {}
call_function  amax                    prims.amax.default                  (convert_element_type, [1])           {}
call_function  broadcast_in_dim        prims.broadcast_in_dim.default      (amax, [3, 1], [0])                   {}
call_function  broadcast_in_dim_1      prims.broadcast_in_dim.default      (broadcast_in_dim, [3, 3], [0, 1])    {}
call_function  sub                     prims.sub.default                   (x_1, broadcast_in_dim_1)             {}
call_function  exp                     prims.exp.default                   (sub,)                                {}
call_function  convert_element_type_1  prims.convert_element_type.default  (exp, torch.float32)                  {}
call_function  sum_1                   prims.sum.default                   (convert_element_type_1, [1])         {}
call_function  broadcast_in_dim_2      prims.broadcast_in_dim.default      (sum_1, [3, 1], [0])                  {}
call_function  broadcast_in_dim_3      prims.broadcast_in_dim.default      (broadcast_in_dim_2, [3, 3], [0, 1])  {}
call_function  div                     prims.div.default                   (exp, broadcast_in_dim_3)             {}
output         output                  output                              (div,)                                {}
```

Commit:
Add ref for `nn.functional.softmax`

Commit:
Add ref for `special.softmax`

nkaretnikov commented:
(will do softmin next)

ngimel commented on torch/_refs/__init__.py:
```diff
@@ -2775,22 +2776,25 @@ def stack(tensors: TensorSequenceType, dim: int = 0) -> TensorLikeType:
 @out_wrapper()
+@elementwise_type_promotion_wrapper(
+    type_promoting_args=("a"),
+    type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.INT_TO_FLOAT,
```
I don't see why you need a type promotion wrapper here as opposed to previous casting to computation dtype. Besides, it's just wrong - softmax doesn't do int_to_float type promotion, it errors out for integer args.

nkaretnikov replied:
i assumed it would be cleaner this way. `int to float` is due to using `div` (as in `addcdiv`), so we'd get the correct(?) behavior if ints are supported later. _i'll change back to explicit cast to computation dtype tho_

nkaretnikov replied:
pushed a version that can use `dtype` before converting to comp dtype and when converting to result dtype

ngimel commented on torch/_refs/__init__.py:
```diff
@@ -2775,22 +2776,25 @@ def stack(tensors: TensorSequenceType, dim: int = 0) -> TensorLikeType:
+    m = torch.amax(a, dim, keepdim=True)
+    e = torch.exp(a - m)
+    r = e / torch.sum(e, dim, keepdim=True)
+    return _maybe_convert_to_dtype(r, result_dtype)  # type: ignore[return-value]
```
`dtype` arg wasn't correctly handled previously, and is still not correctly handled. If `dtype` is supplied, `a` is first converted to dtype.

nkaretnikov replied:
pushed a version that can use dtype before converting to comp dtype and when converting to result dtype

ngimel commented on torch/_refs/nn/functional/__init__.py:
```diff
@@ -233,6 +235,38 @@ def selu(a: TensorLikeType, inplace: bool = False) -> TensorLikeType:
+
+# CompositeImplicitAutograd - don't register decomp
+@out_wrapper()
+@elementwise_type_promotion_wrapper(
```
no type promotion wrapper here, you are calling a ref that handles type promotion

ngimel commented on torch/_refs/special/__init__.py:
```diff
@@ -60,6 +61,20 @@ def logit(self: TensorLikeType, eps: Optional[float] = None) -> TensorLikeType:
 
+# CompositeImplicitAutograd - don't register decomp
+@out_wrapper()
+@elementwise_type_promotion_wrapper(
```
no type promotion wrapper here

Commit:
In `log_softmax` ref, use `dtype` in `get_computation_dtype`

Commit:
Remove keyword-only argument from `log_softmax` ref

For compat with ATen and Python frontend signatures.  See the comment in
`native_functions.yaml`: "log_softmax allows positional dtype, unlike most
operators, because kwonly is BC-breaking when loading jit models."

Commit:
Make sure `log_softmax` decomposes into prim calls

```
import torch
from torch._prims.context import TorchRefsMode
from torch.fx.experimental.proxy_tensor import make_fx

a = torch.randn(3, 3, device="cuda", dtype=torch.float32)

with TorchRefsMode(strict=True):
    make_fx(lambda x: torch.log_softmax(x, 1))(a).graph.print_tabular()
```

Output:

```
opcode         name                    target                              args                           kwargs
-------------  ----------------------  ----------------------------------  -----------------------------  --------
placeholder    x_1                     x_1                                 ()                             {}
call_function  convert_element_type    prims.convert_element_type.default  (x_1, torch.float32)           {}
call_function  amax                    prims.amax.default                  (convert_element_type, [1])    {}
call_function  broadcast_in_dim        prims.broadcast_in_dim.default      (amax, [3, 1], [0])            {}
call_function  abs_1                   prims.abs.default                   (broadcast_in_dim,)            {}
call_function  eq                      prims.eq.default                    (abs_1, inf)                   {}
call_function  where                   prims.where.default                 (eq, 0.0, broadcast_in_dim)    {}
call_function  broadcast_in_dim_1      prims.broadcast_in_dim.default      (where, [3, 3], [0, 1])        {}
call_function  sub                     prims.sub.default                   (x_1, broadcast_in_dim_1)      {}
call_function  exp                     prims.exp.default                   (sub,)                         {}
call_function  convert_element_type_1  prims.convert_element_type.default  (exp, torch.float32)           {}
call_function  sum_1                   prims.sum.default                   (convert_element_type_1, [1])  {}
call_function  broadcast_in_dim_2      prims.broadcast_in_dim.default      (sum_1, [3, 1], [0])           {}
call_function  log                     prims.log.default                   (broadcast_in_dim_2,)          {}
call_function  add                     prims.add.default                   (log, where)                   {}
call_function  broadcast_in_dim_3      prims.broadcast_in_dim.default      (add, [3, 3], [0, 1])          {}
call_function  sub_1                   prims.sub.default                   (x_1, broadcast_in_dim_3)      {}
output         output                  output                              (sub_1,)                       {}
```

Commit:
Add ref for `nn.functional.log_softmax`

Commit:
Add ref for `special.log_softmax`

Commit:
Add ref for `nn.functional.softmin`

nkaretnikov commented:
@pytorchbot rebase

nkaretnikov commented:
@Lezcano ptal

Lezcano commented on torch/_refs/nn/functional/__init__.py:
```diff
@@ -233,6 +237,46 @@ def selu(a: TensorLikeType, inplace: bool = False) -> TensorLikeType:
 
+# copy from torch.nn.functional
+def _get_softmax_dim(name: str, ndim: int, stacklevel: int) -> int:
+    warnings.warn(
+        "Implicit dimension choice for {} has been deprecated. "
+        "Change the call to include dim=X as an argument.".format(name),
+        stacklevel=stacklevel,
+    )
```
We should probably just error out in these cases that have been deprecated.

Lezcano replied:
Also, we should remove the `_stacklevel` arg, as it's a private argument.

nkaretnikov replied:
so, just remove the underscore arg from the refs, correct?

Lezcano replied:
yep, and the warning and all that.

Lezcano replied:
In general, it's fine for refs not to implement behaviour that's been deprecated. We want refs to be forward-looking.

lezcano reviewed (commented):
A few minor things

Lezcano commented on torch/_prims_common/__init__.py:
```diff
@@ -70,6 +70,8 @@ def getnvFuserDtype(dtype: Union[torch.dtype, NumberTypeType]):
     torch.Tensor.__format__,
     torch.Tensor.__repr__,
     torch.Tensor.requires_grad.__get__,  # type: ignore[attr-defined]
+    torch.Tensor.is_contiguous,
+    torch.Tensor.is_quantized.__get__,  # type: ignore[attr-defined]
```
But where are these functions used in this PR?

nkaretnikov replied:
outside of the fx test in the commit message, i don't think these are need. i'll remove, someone can re-add later if necessary

Commit:
Remove items from `torch_function_passthrough` not used in tests

Commit:
[BC breaking] Do not support deprecated behavior in refs

nkaretnikov commented:
@pytorchbot rebase

IvanYashchuk commented:
> i feel like we need a test for all refs that runs under TorchRefsMode(strict=True) to check for this. (thanks to ivan y. for test code)

There are tests already with `strict=True`: https://github.com/pytorch/pytorch/blob/6d6e04d6cc9a3d7cf5d9a2eda5baafd5c3ee75c0/test/test_ops.py#L325
This test verifies that the content of the implementation decomposes down to `prims`, it doesn't check that `torch.softmax` is mapped to `torch._refs.softmax`. The problem would be revealed if some other function from refs would attempt to use `torch.softmax`.

IvanYashchuk commented on torch/_refs/__init__.py:
```diff
@@ -620,11 +622,10 @@ def log10(a):
 def log_softmax(
     a: TensorLikeType,
     dim: int,
-    *,
```
How did you discover this?

The dtype can actually be passed as a positional argument:
https://github.com/pytorch/pytorch/blob/d2b8b8f29121ed23e2b39446d09bba3a7eb96684/aten/src/ATen/native/native_functions.yaml#L3116

But `PythonRefInfo("_refs.log_softmax", ...)` picks up the OpInfo variant without positional dtype and this becomes untested. There two OpInfo for log_softmax:
https://github.com/pytorch/pytorch/blob/d2b8b8f29121ed23e2b39446d09bba3a7eb96684/torch/testing/_internal/common_methods_invocations.py#L14703-L14719

PythonRefInfo("_refs.log_softmax", ...) should use the variant with `partial(sample_inputs_softmax_variant, with_dtype=True)`

nkaretnikov replied:
> How did you discover this?

maybe i'm misunderstanding the comment in the yaml file, then (i quoted it as well in the commit message). my reading of it was "do not use the star here before dtype unlike in other funcs", which is why i removed it from the refs (ditto for softmax). is this not what it says?

IvanYashchuk replied:
And the next step would be to make sure it's properly tested.

Lezcano commented on torch/_refs/nn/functional/__init__.py:
```diff
@@ -233,6 +236,26 @@ def selu(a: TensorLikeType, inplace: bool = False) -> TensorLikeType:
+    dtype: Optional[torch.dtype] = None,
+) -> TensorLikeType:
+    check(dim is not None, lambda: "implicit dim not supported, use dim=X")
+    return torch._refs.softmax(a=a, dim=dim, dtype=dtype)
```
Why are you calling `torch._refs` rather than `torch.softmax` directly?

IvanYashchuk commented on torch/_refs/nn/functional/__init__.py:
```diff
@@ -233,6 +236,26 @@ def selu(a: TensorLikeType, inplace: bool = False) -> TensorLikeType:
 
 
+# CompositeImplicitAutograd - don't register decomp
+def softmax(
+    a: TensorLikeType,
+    dim: Optional[int] = None,
+    dtype: Optional[torch.dtype] = None,
+) -> TensorLikeType:
```
Missing `_stacklevel` argument.

```py
import torch
from torch._prims.context import TorchRefsMode
a = torch.randn(3, 3, device="cuda", dtype=torch.float32)
with TorchRefsMode(strict=True):
    torch.nn.functional.softmax(a, 1)
```
```py
TypeError: softmax() got an unexpected keyword argument '_stacklevel'
```

Lezcano replied:
I suggested removing it here:
https://github.com/pytorch/pytorch/pull/84502#discussion_r962989250
Do we need to keep these deprecation arguments in the refs? :/

IvanYashchuk replied:
Yes, as you see there's no way to use `torch.nn.functional.softmax` under TorchRefsMode without this argument.

nkaretnikov replied:
added a test for this behavior in "Test aliases when testing refs". if it's too much for this PR, i can just revert it later, once we see that ci passes. as of now, aliases are not tested within the context afaict

nkaretnikov replied:
upd: mario said it's okay to keep these alias tests. the only thing i'm concerned about is that there'll be a bit of duplication during testing. eg if you have separate refinfos for functional softmax, special softmax, and softmax, all those end up tested against the same softmax opinfo, so all aliases are tested for each of those as well. but perhaps it's fine(?) because doing some fancy filtering of those might be error-prone

Lezcano replied:
How could you have separate refinfos for softmax and special softmax? We shouldn't allow to register refinfos for aliases if we can avoid it.

If you think we could write a test for that, that could go in another PR.

Commit:
Call functions from the `torch` namespace

Commit:
Use `with_dtype` when testing refs

Commit:
Add back `_stacklevel` for compat with `TorchRefsMode(strict=True)`

Commit:
Test aliases when testing refs

nkaretnikov commented:
@pytorchbot rebase

Lezcano commented on torch/_refs/nn/functional/__init__.py:
```diff
@@ -357,6 +382,17 @@ def l1_loss(
+def log_softmax(
+    a: TensorLikeType,
+    dim: Optional[int] = None,
+    _stacklevel: int = 3,  # for compat when using TorchRefsMode(strict=True)
```
Thank you for the comment!

lezcano reviewed (approved):
Cool! And thank you for the enhanced tests! They are going to be quite useful going forward.

I'd like to have @IvanYashchuk have a look at this one as well, see what he thinks.

Commit:
Use functional style instead of allocating a list

Commit:
Fix alignment

Commit:
Allow to skip alias tests (enabled by default)

IvanYashchuk reviewed (approved):
Thank you for improving our tests!

nkaretnikov commented:
@ngimel ci is green, ptal

summary:
- improve/add `softmax`, `log_softmax`, `softmin`, add aliases, use `dtype` arg in comp dtype, test with dtype
- add `test_aliases` to `OpInfo` (enabled by default) to test `OpInfo` aliases in ref tests

mruberry commented on torch/_refs/nn/functional/__init__.py:
```diff
@@ -357,6 +382,17 @@ def l1_loss(
+    _stacklevel: int = 3,  # for compat when using TorchRefsMode(strict=True)
+    dtype: Optional[torch.dtype] = None,
+) -> TensorLikeType:
+    check(dim is not None, lambda: "implicit dim not supported, use dim=X")
```
Let's add a comment for these checks that explains the behavior is deprecated in regular PyTorch, and this error is to inform users how to update their calls (otherwise it seems like we would just not put a default value for `dim` in the parameters?)

mruberry commented on torch/testing/_internal/opinfo/core.py:
```diff
@@ -793,6 +793,9 @@ class OpInfo(object):
 
     test_neg_view: bool = True
 
+    # test aliases in '_ref_test_helper'
```
Maybe "Whether to test aliases of the operator (like arccosh for acosh)"?

mruberry commented on test/test_ops.py:
```diff
@@ -316,7 +327,7 @@ def _distance(a, b):
     # Tests that experimental Python References perform the same computation
     # as the operators they reference, when operator calls in the torch
-    # namesapce are remapped to the refs namespace (torch.foo becomes refs.foo).
+    # namespace are remapped to the refs namespace (torch.foo becomes refs.foo).
```
Nice spelling fix

mruberry reviewed (commented):
Overall looks good, @nkaretnikov -- I added some questions and and a few minor suggestions inline for your review. Looking forward to hearing your thoughts!

Commit:
Add comments explaining the error for implicit dim

nkaretnikov commented on torch/_refs/special/__init__.py:
```diff
@@ -60,6 +61,24 @@ def logit(self: TensorLikeType, eps: Optional[float] = None) -> TensorLikeType:
 
 
+# CompositeImplicitAutograd - don't register decomp
+def log_softmax(
```
no, i think. `special` fns don't support `out`, so this will break `out` tests (which currently expect `out` arg to be unsupported)

Commit:
Update the comment about `test_aliases`

nkaretnikov commented on test/test_ops.py:
```diff
@@ -227,17 +227,26 @@ def _ref_test_helper(self, ctx, device, dtype, op, skip_zero_numel=False, skip_z
+            # because 'op.aliases' is empty.  When the aliases are executed
+            # within the context, they will behave as refs.  This is done to
+            # check for compatibility between all APIs.
+            aliases = op.torch_opinfo.aliases if op.test_aliases else ()
```
doing `bar is foo` for refs might be tricky because of api incompat. see the example in the other comment for `special`. also feels less transparent for the users because it would rely on object ids.

i think it would be easier to have:
- the above alias test and
- to avoid duplicate work in the alias test, make sure users cannot register aliases when declaring refs in the `PythonRefInfo` constructor (https://github.com/pytorch/pytorch/pull/84502#discussion_r964717129)

  example: instead of having refs for `softmax`, `special.softmax`, and `nn.functional.softmax`, you'll just have one ref for `softmax` testing all the aliases.
  
the current plan is to have follow up PRs that would:
  - disallow declaring aliases as refs
  - maybe: add alias testing to tests where we'd lose coverage because of the above change (decomps?)
  - add missing refs for aliases (exposed by the current alias test, where `test_aliases=False` atm).
  
  wdyt?

nkaretnikov replied:
after thinking more about this, i think this might be better:
- i'll revert the alias testing part of this in the PR such that only operations (and aliases) are added here (i should have used ghstack for this, i know)
- i'll resubmit the alias testing part with the other testing-infra-related changes such that we can discuss and merge them all at once (instead of relying on follow up commits)

mruberry replied:
Separating the aliasing part makes sense to me conceptually, I'll focus on reviewing the references

nkaretnikov commented:
@pytorchbot rebase -s

Commit:
Disallow aliases in `python_ref_db`

nkaretnikov commented:
@pytorchbot rebase -s

nkaretnikov commented:
Apologies for the noise, this turned out quite messy! I've explored various options over the weekend. I think this is the best way to do it (as implemented in this branch right now):
- we keep the test against all the aliases in `TestRefsOpsInfo` to catch API compat issues
- we disallow aliases in `python_ref_db` (entries with `*PythonRefInfo`), a new test added for this
- we don't lose coverage because the current ref/decomp tests sample from `__all__` rather than from `python_ref_db`
- we can sample from either `__all__` or `python_ref_db` in any other tests too if needed
- we don't remove any entries not added by this branch (I've only had to make minor changes to `logit` due to how it was defined)
- this still allows for duplicate refs such as those having variants (`div`) or nvfuser-specific ones (`ops.nvprims.var_mean`)

Lezcano commented on test/test_ops.py:
```diff
@@ -1664,7 +1696,15 @@ class TestRefsOpsInfo(TestCase):
+
+    @parametrize("op", ref_db_names)
+    def test_ref_db_has_no_aliases(self, op):
+        self.assertTrue(op not in self.ref_alias_names_set,
+                        f"alias {op} in python_ref_db, which is not allowed")
```
Can you improve this message?

nkaretnikov replied:
could you suggest one? i had tried some before i ended up on this one

Lezcano commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -16262,15 +16289,11 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
     #
     # Elementwise Unary Special OpInfos
     #
-    ElementwiseUnaryPythonRefInfo(
-        "_refs.special.logit",
-        torch_opinfo_name="logit",
-        supports_nvfuser=False,
-    ),
```
Shouldn't this one be the OG OpInfo, and then have `torch.logit` tested in terms of this?
This is a bit confusing, as it's not clear which one is the implementation and which one is the alias. At the moment, the implementation is in `refs.special.logit`, while the OpInfo is that of `torch.logit`.

lezcano reviewed (commented):
This looks mostly good. Just left two comments.

Commit:
Improve error message

Commit:
Swap the alias and op in `logit` `OpInfo`

This properly reflects which one is the alias.

Commit:
[functorch] Adapt to changes in `logit` `OpInfo`

mruberry commented on torch/_refs/nn/functional/__init__.py:
```diff
@@ -233,6 +236,36 @@ def selu(a: TensorLikeType, inplace: bool = False) -> TensorLikeType:
+    _stacklevel: int = 3,  # for compat when using TorchRefsMode(strict=True)
+    dtype: Optional[torch.dtype] = None,
+) -> TensorLikeType:
+    # This is for compat with regular PyTorch, which has this behavior
```
Nit: Instead of saying "this" the comment should explicitly specify what it's referring to

mruberry reviewed (commented):
The new references and comments look good, I made one small nit. Ping me when the PR is separated into the aliases part and the references part!

Commit:
[JIT] Adapt to changes in `logit` `OpInfo`

nkaretnikov commented:
resubmitted as a stack here https://github.com/pytorch/pytorch/pull/84956, closing
